design/authenticators/application-identity/application-identity.md

```diff
@@ -0,0 +1,466 @@
+
+#### Validate the Application Identity matches the authentication request
+
+Once we extracted the Application Identity from the host's annotations, and
+verified it is defined correctly, we can validate that the authentication
+request is valid by validating the Application Identity. For example, if
+the k8s host is permitted to authenticate from namespace `namespace-1`, and
+the authentication request was sent from a pod that is in namespace
+`namespace-2` we will fail the request.
+
+In a way, we can define this step as comparing 2 objects of Application
+Identity - one that was extracted from the host and one that was extracted
+from the authentication request. In the example above we will have the
+following Application Identity objects:
+ - Host Application Identity
+   - constraints
+     - namespace: `namespace-1`
+ - Authentication request Application Identity
+   - constraints
+     - namespace: `namespace-2`
+     
+All we need to do is iterate over the Authentication request Application Identity
+`constraints` hash and verify that all of its fields have the same values as the 
+corresponding `constraints` hash fields in the Host Application Identity object.
+  
+In this case the general class will be defined as follows:
+- input
+  - Host Application Identity object
+  - Authentication request Application Identity object
+- output
+  - Nothing if validation succeeded, InvalidApplicationIdentity error if it
+   failed
+   
+### Implementation
+
+In conclusion, our general ValidateApplication class will need the following
+inputs:
+  - Authenticator name (e.g 'authn-azure')
+  - Authenticator's service-id (e.g 'prod')
+  - host's annotations
+  - Different lists of annotation names
+    - permitted
+    - required
+    - non-permitted combinations
+  - Authentication request Application Identity object
+  
+It will perform the following steps:
+  1. Build the Host Application Identity object using the:
+     - authenticator name
+     - authenticator service-id
+     - host's annotations
+     - list of permitted annotations
+  2. Verify the Host Application Identity is configured correctly using the:
+     - Host Application Identity object built in the previous step
+     - list of required constraints
+     - list of non-permitted constraints combinations
+  3. Verify the Host Application Identity matches the authentication request
+   using the:
+     - Host Application Identity object built in the first step
+     - Authentication request Application Identity object given in the input
+      to this class
+      
+Let's see how the classes will look, at a high level.
+
+#### Application Identity class
+
+We will have 2 different Application Identity classes. Although their
+interface is the same, they are not entirely identical. 
+
+At first, the `constraints` method will return a simple key-value hash. We
+should challenge that and find a better way to do it in Ruby.
```

Comment:
@jonahx any idea?

Reply:
The dangerous way to do this is with `method_missing` (powerful, but treat with great care):
```
➜  dap-wiki git:(master) irb
2.5.1 :001 > class Foo
2.5.1 :002?>     def initialize()
2.5.1 :003?>         @attributes = {
2.5.1 :004 >               bing: 'bang',
2.5.1 :005 >               ping: 'pong'
2.5.1 :006?>           }
2.5.1 :007?>       end
2.5.1 :008?>
2.5.1 :009?>     def method_missing(m, *args, &blk)
2.5.1 :010?>         if @attributes.key?(m.to_sym)
2.5.1 :011?>             @attributes[m.to_sym]
2.5.1 :012?>           else
2.5.1 :013?>             super
2.5.1 :014?>           end
2.5.1 :015?>       end
2.5.1 :016?>   end
 => :method_missing
2.5.1 :017 > foo = Foo.new
 => #<Foo:0x00007f9e74186f40 @attributes={:bing=>"bang", :ping=>"pong"}>
2.5.1 :018 > foo.ping
 => "pong"
2.5.1 :019 > foo.baz
Traceback (most recent call last):
        3: from /Users/jasonvanderhoof/.rvm/rubies/ruby-2.5.1/bin/irb:11:in `<main>'
        2: from (irb):19
        1: from (irb):13:in `method_missing'
NoMethodError (undefined method `baz' for #<Foo:0x00007f9e74186f40>)
```